haxe/ui/core/Component.hx

```diff
@@ -1746,74 +1732,285 @@ class Component extends ComponentBase implements IComponentBase implements IClon
+    private var _isAllInvalid:Bool = false;
+    private var _isValidating:Bool = false;
+    private var _isDisposed:Bool = false;
+    private var _invalidateCount:Int = 0;
+
+    private var _depth:Int;
+    @:dox(hide)
+    public var depth(get, set):Int;
+    private function get_depth():Int {
+        return _depth;
+    }
+    private function set_depth(value:Int):Int {
+        if (_depth == value) {
+            return value;
+        }
+
+        _depth = value;
+
+        if (_children != null) {
+            var childDepth = (_depth != -1) ? _depth + 1 : -1;
+            for(c in _children) {
+                c.depth = childDepth;
+            }
+        }
+
+        return value;
+    }
+
     /**
-     Invalidate this components layout, may result in multiple calls to `invalidateDisplay` and `invalidateLayout` of its children
+     This method validates the tasks pending in the component.
     **/
     @:dox(group = "Invalidation related properties and methods")
-    public function invalidateLayout() {
-        if (_ready == false || _layout == null) {
+    public function validate():Void {
+        if(_ready == false
+            || _isDisposed == true      //we don't want to validate disposed components, but they may have been left in the queue.
+            ||_isValidating == true     //we were already validating, the existing validation will continue.
+            || isInvalid() == false) {  //if none is invalid, exit.
             return;
         }
 
-        if ((_layoutInvalidating == true || _layoutLocked == true) && _layoutReinvalidation == false) {
-            // means that if a request to invalidate comes through and were busy
-            // (like async resources), we make note to invalidate when were done
-            _layoutReinvalidation = true;
-            return;
+        _isValidating = true;
+
+        validateInternal();
+
+        for (flag in _invalidationFlags.keys()) {
+            _invalidationFlags.remove(flag);
+        }
+
+        _isAllInvalid = false;
+
+        for (flag in _delayedInvalidationFlags.keys()) {
+            if (flag == InvalidationFlags.ALL) {
+                _isAllInvalid = true;
+            } else {
+                _invalidationFlags.set(flag, true);
+            }
+            _delayedInvalidationFlags.remove(flag);
+        }
+        _isValidating = false;
+    }
+
+    private function validateInternal():Void {
+        var styleInvalid = isInvalid(InvalidationFlags.STYLE);
+        var positionInvalid = isInvalid(InvalidationFlags.POSITION);
+        var displayInvalid = isInvalid(InvalidationFlags.DISPLAY);
+        var layoutInvalid = isInvalid(InvalidationFlags.LAYOUT) && _layoutLocked == false;
+
+        if (styleInvalid) {
+            validateStyle();
+        }
+
+        if (hasTextDisplay()) {
+            getTextDisplay().validate();
+        }
+
+        if (hasTextInput()) {
+            getTextInput().validate();
+        }
+
+        //TODO
```

Comment:
It works. But it can be optimized with the new flow. The performance test in the PR is button with label and icon.